### PR TITLE
Multi version lsm docs

### LINKED ISSUE #7841: add documentation for multi-version lsm
1. write out an example of how to use it
  - in the model
  - how to export to the catalogue (same as before)
  - how to upgrade instances (add example python script e.g.?)
  - how to add/remove new versions
  - difference between v1 and v2 endpoints
  - effect of updating the default version on the api
2. put the example code in both the lsm module and core repo https://github.com/inmanta/inmanta-core/tree/master/docs/lsm and https://github.com/inmanta/lsm/tree/master/tests/examples_docs
3. add a testcase to the example in lsm module
4. update the sync check https://github.com/inmanta/lsm/blob/master/tests/lsm_documentation_diff.txt

### PR DESCRIPTION
**Pull request by jptrindade**
# Description

The first draft of multi version docs. Might still be lacking but wanted to have this out for review before I leave for vacation.

From the ticket description:

    write out an example of how to use it

 - [x]  in the model
 - [ ]  how to export to the catalogue (same as before)
 - [x]  how to upgrade instances (add example python script e.g.?)
 - [x]  how to add/remove new versions
 - [x]  difference between v1 and v2 endpoints
 - [x]  effect of updating the default version on the api
 - [x] put the example code in both the lsm module and core repo https://github.com/inmanta/inmanta-core/tree/master/docs/lsm and https://github.com/inmanta/lsm/tree/master/tests/examples_docs
 - [x] add a testcase to the example in lsm module
 - [x]  update the sync check https://github.com/inmanta/lsm/blob/master/tests/lsm_documentation_diff.txt

closes #7841 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: docs/lsm/multi_version/multi_version.rst [hunk 1]
```diff
@@ -0,0 +1,92 @@
+**************
+Multi-version
+**************
+
+The goal of multi-version lsm is to have versioning as a way to introduce changes to service entities while
+supporting the older behaviour and remaining backwards compatible.
+Each service entity will now have versions, which can introduce breaking changes while maintaining support for older
+versions.
+
+Why do we want to use  multi-version LSM and not different services?
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:10]:
```suggestion
Why use multi-version LSM?
```

File: docs/lsm/multi_version/multi_version.rst [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+====================================================================
+A service represents a high level desired state. With a new version we want to introduce some changes to a service
+(i.e. add a new option, or change how some minor things are done), but overall, keep the same intent.
+We want to be able to let a service evolve over time while maintaining backwards compatibility and be able to upgrade
+from one version to another without drastically changing the intent of the service.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:15]: I would rephrase this to a very practical `when to use multi-version lsm`  section. I think you want to use when 1. you want to offer multiple api schema versions to the same service or 2. you want to upgrade a service in a way that is not supported by the automated upgrade mechanism (if possible, you can link to the docs that describe what is supported, if we have such docs)

File: docs/lsm/multi_version/multi_version.rst [hunk 3]
```diff
@@ -0,0 +1,92 @@
+**************
+Multi-version
+**************
+
+The goal of multi-version lsm is to have versioning as a way to introduce changes to service entities while
+supporting the older behaviour and remaining backwards compatible.
+Each service entity will now have versions, which can introduce breaking changes while maintaining support for older
+versions.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:8]:
```suggestion
Multi-version lsm allows you to have multiple api versions for the same service. 
```

File: docs/lsm/multi_version/multi_version.rst [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+Every service entity now has a version. The existing service entities and each new one created with the
+``lsm::ServiceEntityBinding`` or ``lsm::ServiceEntityBindingV2`` will have one version numbered 0 (which will be the
+default version for this service).
+
+Additionally we can provide a default version to ``lsm::ServiceBinding`` to change which version will be accessible
+through the v1 endpoints.
+
+Updating a service entity version is not supported for versioned entities. The new workflow is to create a new version
+with the required changes. However, updates to a version 0 of a service entity are allowed, if it is the only version
+of that service.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:32]: This is too lsm developer oriented. Backwards compatibility is important to us, probably less to our end-user. 

I think you have to start with explaining which api calls can be used to manipulate multi version instances, then explain that the same endpoints exists without a version in the url and these use the default version.

File: docs/lsm/multi_version/multi_version.rst [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+Migrating from an existing unversioned model
+============================================
+
+Using the :ref:`model to create InterfaceIPAssignments <intro_example>` as an example, we will see how we can turn this
+entity into a versioned one.
+
+We just need to change our existing ``lsm::ServiceEntityBinding`` into a ``lsm::ServiceBinding``. Set the
+``default_version`` to 0 and create a ``lsm::ServiceBindingVersion`` with the information that we had on our previous
+binding.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:43]: I would make this a bulleted list or numbered list

File: docs/lsm/multi_version/multi_version.rst [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+Additionally we can provide a default version to ``lsm::ServiceBinding`` to change which version will be accessible
+through the v1 endpoints.
+
+Updating a service entity version is not supported for versioned entities. The new workflow is to create a new version
+with the required changes. However, updates to a version 0 of a service entity are allowed, if it is the only version
+of that service.
+
+
+Migrating from an existing unversioned model
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:35]: I think this should go before the previous section, so the example is immediately after the when to use

Comment (wouterdb) [reply docs/lsm/multi_version/multi_version.rst:35]:
```suggestion
Using multi version lsm
```

**Review by wouterdb — CHANGES_REQUESTED** [review]
nice

**Comment by jptrindade** [comment]
Should I also add here that we still do not have support for partial compile and inter-service relations?

File: docs/lsm/multi_version/multi_version.rst (hunk 4 above)
Comment (jptrindade) [reply docs/lsm/multi_version/multi_version.rst]: Made a section for the API calls. Let me know if you want me to go more in-depth in that section

File: docs/lsm/multi_version/multi_version.rst [hunk 7]
```diff
@@ -0,0 +1,99 @@
+**************
+Multi-version
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:2]:
```suggestion
Multi-version LSM
```

File: docs/lsm/multi_version/multi_version.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+and recompile and export the new model.
+
+NOTE: We cannot remove service entity versions that have active instances.
+
+API endpoints
+=======================
+The following API endpoints were added in order to manage versioned services:
+
+* `GET lsm/v2/service_catalog` : List all service entity versions of each defined services entity type in the service catalog
+* `POST lsm/v2/service_catalog`: Create a new service entity type in the service catalog and assign a default version to it.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:61]: end-users will never use this directly, not relevant here
```suggestion
```

File: docs/lsm/multi_version/multi_version.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+* `GET lsm/v2/service_catalog/<service_entity>/<version>/schema`: Get the json schema for a service entity version.
+* `GET lsm/v2/service_catalog/<service_entity>`: Get all versions of the service entity type from the service catalog.
+* `GET lsm/v2/service_catalog/<service_entity>/<version>`: Get one service entity version from the service catalog.
+* `PATCH lsm/v2/service_catalog/<service_entity>`: Update an existing service entity type in the service catalog or add a new one.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:65]: end-users will never use this directly, not relevant here
```suggestion
```

File: docs/lsm/multi_version/multi_version.rst [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+* `DELETE lsm/v2/service_catalog/<service_entity>/<version>`: Delete an existing service entity version from the service catalog.
+* `GET lsm/v2/service_catalog/<service_entity>/<version>/config`: Get the config settings for a service entity version.
+* `POST lsm/v2/service_catalog/<service_entity>/<version>/config`: Set the config for a service entity version.
+* `PATCH lsm/v1/service_inventory/<service_entity>/<service_id>/update_entity_version`: Migrate a service instance from one service entity version to another.
+
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:70]: I miss the ones to interact with the instances.

File: docs/lsm/multi_version/multi_version.rst [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+The older v1 endpoints are still supported.
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:71]:
```suggestion
The  v1 endpoints are still supported.
```

File: docs/lsm/multi_version/multi_version.rst [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+or by calling the `lsm_services_update_entity_version` API method on the Inmanta client.
+
+To do this we need to provide all 3 attribute sets that we want the instance to have on the new entity version.
+These attribute sets are validated against the schema of the new version.
+We also need to provide the target state that we want to set the instance to.
+
+NOTE: This change is impossible to rollback since we override each attribute set. And each attribute set needs to be
+compatible with the target entity version.
+
+
```
Comment (wouterdb) [inline docs/lsm/multi_version/multi_version.rst:94]: I would add a code example here as well (short python script e.g.)

**Review by wouterdb — CHANGES_REQUESTED** [review]
Looks nice!

File: docs/lsm/multi_version/multi_version.rst (hunk 10 above)
Comment (jptrindade) [reply docs/lsm/multi_version/multi_version.rst:68]: There are no new endpoints to interact with instances. When applicable we just have an optional `service_entity_version` argument. I'll add this information here as well.

**Review by wouterdb — APPROVED** [review]
